hail/src/main/scala/is/hail/io/fs/GoogleStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+        var i = off
+        var remaining = len
+        while (remaining > 0) {
+          if (bb.remaining() == 0)
+            flush()
+          val toTransfer = math.min(bb.remaining(), remaining)
+          bb.put(bytes, i, toTransfer)
+          i += toTransfer
+          remaining -= toTransfer
+          pos += toTransfer
+        }
+      }
+
+      override def close(): Unit = {
+        flush()
+        write.close()
+      }
+
+      def getPosition: Long = pos
+    }
+
+    new WrappedPositionedDataOutputStream(os)
+  }
+
+  def mkDir(dirname: String): Unit = ()
+
+  def delete(filename: String, recursive: Boolean): Unit = {
+    val (bucket, path) = getBucketPath(filename)
+    if (recursive) {
+      val it = storage.list(bucket, BlobListOption.prefix(path))
+        .getValues.iterator.asScala
+      while (it.hasNext) {
+        storage.delete(it.next().getBlobId)
+      }
+    } else
+      storage.delete(bucket, path)
+  }
+
+  def glob(filename: String): Array[FileStatus] = {
+    var (bucket, path) = getBucketPath(filename)
+    path = dropTrailingSlash(path)
+
+    val components =
+      if (path == "")
+        Array.empty[String]
+      else
+        path.split("/")
+
+    val javaFS = FileSystems.getDefault
+
+    val ab = new mutable.ArrayBuffer[FileStatus]()
+    def f(prefix: String, fs: FileStatus, i: Int): Unit = {
+      assert(!prefix.endsWith("/"), prefix)
+
+      if (i == components.length) {
+        var t = fs
+        if (t == null) {
+          try {
+            t = fileStatus(prefix)
+          } catch {
+            case _: FileNotFoundException =>
+          }
+        }
+        if (t != null)
+          ab += t
+      }
+
+      if (i < components.length) {
+        val c = components(i)
+        if (containsWildcard(c)) {
+          val m = javaFS.getPathMatcher(s"glob:$c")
+          for (cfs <- listStatus(prefix)) {
+            val p = dropTrailingSlash(cfs.getPath)
+            val d = p.drop(prefix.length + 1)
+            if (m.matches(javaFS.getPath(d))) {
+              f(p, cfs, i + 1)
+            }
+          }
+        } else
+          f(s"$prefix/$c", null, i + 1)
+      }
+    }
+
+    f(s"gs://$bucket", null, 0)
+    ab.toArray
+  }
+
+  def globAll(filenames: Iterable[String]): Array[String] =
+    globAllStatuses(filenames).map(_.getPath)
+
+  def globAllStatuses(filenames: Iterable[String]): Array[FileStatus] = filenames.flatMap(glob).toArray
+
+  def listStatus(filename: String): Array[FileStatus] = {
+    var (bucket, path) = getBucketPath(filename)
+    if (!path.endsWith("/"))
+      path = path + "/"
+
+    val blobs = storage.list(bucket, BlobListOption.prefix(path), BlobListOption.currentDirectory())
+
+    blobs.getValues.iterator.asScala
+      .map(b => GoogleStorageFileStatus(b))
+      .toArray
+  }
+
+  def fileStatus(filename: String): FileStatus = {
+    var (bucket, path) = getBucketPath(filename)
+
+    while (path.endsWith("/"))
+      path = path.dropRight(1)
```

Comment:
Can you ever have > 1 slash here? If so dropTrailingSlash.

Reply:
This is a while, so this handles > 1 slash.  I switched to dropTrainingSlash, thanks.